### PR TITLE
Alert user when the action could lead to the restart of the notebooks

### LINKED ISSUE #847: [Feature Request]: Warn users of pod restart when editing workbench
### Feature description

When editing a workbench most changes require the pod to be restarted if it is currently running.  Before saving the changes a warning should appear stating that the pod will need to be restarted and users should save any notebook data before proceeding.

Restarting the pod without warning could result in loss of user data that has not been saved or pushed to git.

### Describe alternatives you've considered

Instead of automatically restarting the pod or giving a warning, you could add an extra step after the changes have been saved.   A message could be displayed on the workbench item indicating that the changes will be applied once the pod has been restarted with a button to "Restart Now".

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by DaoDaoNoCode**
## Description

Closes #847 

## How Has This Been Tested?

Check the following page/modals:
- Editing workbench
- Add/Edit storage modal
- Add/Edit data connections modal

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by DaoDaoNoCode** [comment]
Waiting for final reviews from the UX for the wording in the alert.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
I'm thinking we may want to only show it when we know it will happen.

Creation of the Data Connection or PVC when they are not associating to a workbench is a scenario where we can do it without impact.

There is also the concept of a workbench being "off"... that also won't impact it 🤔 

@kywalker-rh thoughts on smart detection? Do you think it adds value?

**Comment by kywalker-rh** [comment]
> 

@andrewballantyne I agree. If we have the ability to only show the alert when it is actually going to be in effect, it will give it more impact, and also allow users to make a better informed choice about how to proceed.

File: frontend/src/pages/projects/notebook/utils.ts
Comment (andrewballantyne) [inline frontend/src/pages/projects/notebook/utils.ts:257]:
```suggestion
export const willNotebooksRestart = (
```

Any reason you didn't make this a hook? It seems unnecessary to request the caller to pass you the `notebookStates` when it's a context variable.

File: frontend/src/pages/projects/screens/detail/storage/ManageStorageModal.tsx
Comment (andrewballantyne) [inline frontend/src/pages/projects/screens/detail/storage/ManageStorageModal.tsx:25]: Nit, order

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Need to test it... initial code feedback

File: frontend/src/pages/projects/components/NotebookRestartAlert.tsx
Comment (andrewballantyne) [inline frontend/src/pages/projects/components/NotebookRestartAlert.tsx:7]: Do we want to list the workbenches? cc @kywalker-rh

File: frontend/src/pages/projects/components/NotebookRestartAlert.tsx
Comment (andrewballantyne) [inline frontend/src/pages/projects/components/NotebookRestartAlert.tsx:6]:
```suggestion
    <Alert variant="info" isInline title="Restarting associated workbenches">
```
🤔  @kywalker-rh  do we know what is more important? The fact that it is restarting or the fact that you could lose data.

What about those moments were we want to tell them we are restarting it _yet again_ in the middle of it starting? Like they add a Storage, add a Data Connection immediately after -- it's likely still restarting because of the new storage.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
I think we need to put it in the sticky footer -- this isn't going to work. It's under the fold / easy to miss at the bottom of the form.

Screenshot 2023-01-20 at 12 53 40 PM
Screenshot 2023-01-20 at 12 53 46 PM

---

Interesting, while it is starting we don't warn 🤔 I think that's risky. But probably not a big deal. Since they are unlikely to submit at the right moment, in the middle of it doing something. Just feels weird in the UI.

I wonder if we can consider starting too... I know they can't really lose unsaved work, but if the Notebook is starting/started, it has sorta the same value of saying "we are going to do this to you -- you will have to wait longer to use your notebook".

File: frontend/src/pages/projects/notebook/utils.ts
Comment (DaoDaoNoCode) [reply frontend/src/pages/projects/notebook/utils.ts:257]: Yeah I made it a hook at first, but when using it, I found some scenarios in which I cannot use a hook... but now when I go back and check I find that the problem is solved. So I will change it to a hook.

File: frontend/src/pages/projects/components/NotebookRestartAlert.tsx
Comment (andrewballantyne) [reply frontend/src/pages/projects/components/NotebookRestartAlert.tsx:7]: - adding a storage directly to a workbench
- adding/removing a storage with a singular workbench either being added or removed from a list -- maybe both?
- adding/removing any number of data connections from workbenches
- directly editing a workbench

Aside from the last one & first one -- I think there is some chance of having multiple impacted. @DaoDaoNoCode do you think we can get the names listed?

Comment (DaoDaoNoCode) [reply frontend/src/pages/projects/components/NotebookRestartAlert.tsx:7]: I think so, it should not be hard to implement.

Comment (kaedward) [reply frontend/src/pages/projects/components/NotebookRestartAlert.tsx:7]: If it's only the one workbench restarting, I'll suggest some new microcopy:
"Updating this workbench will cause it to restart. To avoid losing your work, save any recent data."

^^This also matches the button copy (Update workbench)

LMK if we may need to list multiple workbenches and I can suggest a way to work that into the alert!

Comment (andrewballantyne) [reply frontend/src/pages/projects/components/NotebookRestartAlert.tsx:7]: @kaedward we definitely are looking at needing to list the names of workbenches. The UI doesn't hold onto everything. So you can remove a workbench and add a different workbench. Thus now it impacts two potentially running workbenches. 

Do we want to say "running x, not running y" or only if they are running? There is a lot of ways to go about this. "Impacted workbenches are: x, y, and z" (irrespective of running state).

File: frontend/src/pages/projects/components/NotebookRestartAlert.tsx
Comment (kywalker-rh) [reply frontend/src/pages/projects/components/NotebookRestartAlert.tsx:6]: The unsaved work is the biggest issue here. Restarting might be annoying, but the worst outcome would be restarting and losing your work because it wasn't saved.

Comment (kaedward) [reply frontend/src/pages/projects/components/NotebookRestartAlert.tsx:6]: I would think that, if the workbench is in the middle of restarting, the user won't have made any new changes. So in this case I'd think "Restarting associated workbenches" is more important. @kywalker-rh please correct me if I'm wrong :-)

Also, is "Restarting workbench" more accurate for this alert? Unless multiple workbenches are restarting at the same time, "Restarting associated workbenches" might be confusing.

Comment (kywalker-rh) [reply frontend/src/pages/projects/components/NotebookRestartAlert.tsx:6]: My comment was based on the entire concept of having an alert... if this is specifically talking about alerting users after they have already made changes, and we need to restart again, then yes, I agree with @kaedward, sorry for any confusion

File: frontend/src/pages/projects/components/NotebookRestartAlert.tsx
Comment (kaedward) [reply frontend/src/pages/projects/components/NotebookRestartAlert.tsx:7]: @andrewballantyne Will this still be an inline alert, and will it update as the user selects workbenches to update? And will the list of workbenches be included in that alert, or somewhere else? 

I think listing only the running workbenches should be fine, but I'm open to ideas. My reasoning is that the running ones are the workbenches that the user may want to check + save their work in before proceeding. I think if I got an alert like this I would want to make sure I checked ALL of the listed workbenches, running or not, especially if I skimmed the text and just saw "save your work". 

I'm imagining something like this:
"Running workbenches will restart upon updating. To avoid losing your work, save any recent data in the following running workbenches: X, Y, Z"

File: frontend/src/pages/projects/notebook/useWillNotebooksRestart.ts
Comment (DaoDaoNoCode) [reply frontend/src/pages/projects/notebook/useWillNotebooksRestart.ts:10]: Ah, I was returning the notebook display name list before and changed that, but forgot to change the method to this easier one, thanks for pointing that out.

Comment (andrewballantyne) [reply frontend/src/pages/projects/notebook/useWillNotebooksRestart.ts:10]: FWIW, it is easier to read a `.filter` and a `.map` than a `.reduce`.

- [filter] I want to remove these items
- [map] I want them to be structured like _this_

is easier to understand than

- [reduce] I may reduce, change, or add back items at any point -- they may even be different items all together

**Review by andrewballantyne — APPROVED** [review]
This appears to cover the broad strokes... we may want to revisit this if the interim message changing and flow feels weird as the Workbenches go from Starting to Running. But it conveys the information well -- so the user isn't caught off guard anymore when the Workbench restarts abruptly

**Pull request merged**
